### PR TITLE
Refactor EdgeHitProcessor.processHit to clean up error handling

### PR DESCRIPTION
**Pull request by kevinlind**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: code/edge/src/main/java/com/adobe/marketing/mobile/EdgeHitProcessor.java [hunk 1]
```diff
@@ -276,6 +197,160 @@ public void onComplete() {
...
+				LOG_TAG,
+				LOG_SOURCE,
+				"Failed to build the consent payload, dropping current event (%s).",
+				entity.getEvent().getUniqueIdentifier()
+			);
+
+			return true; // Request complete, don't retry hit
+		}
+
+		Map<String, Object> edgeConfig = entity.getConfiguration();
```
Comment (addb) [inline code/edge/src/main/java/com/adobe/marketing/mobile/EdgeHitProcessor.java:310]: Should we add helper for create edgeEndpoint and reuse and reduce the code in processExperienceEventHit & processUpdateConsentEventHit

File: code/edge/src/main/java/com/adobe/marketing/mobile/EdgeHitProcessor.java [hunk 2]
```diff
@@ -276,6 +197,160 @@ public void onComplete() {
...
+			EdgeConstants.SharedState.Configuration.EDGE_CONFIG_ID,
+			null
+		);
+		if (StringUtils.isNullOrEmpty(edgeConfigId)) {
+			// The Edge configuration ID value should get validated when creating the Hit,
+			// so we shouldn't get here in production.
+			Log.debug(
+				LOG_TAG,
+				LOG_SOURCE,
+				"Cannot process hit as the Edge Network configuration ID is null or empty, dropping current event (%s).",
```
Comment (emdobrin) [inline code/edge/src/main/java/com/adobe/marketing/mobile/EdgeHitProcessor.java:322]: nit: can you tweak these two log messages (experience event and consent) to indicate what kind of hit they are processing

**Review by emdobrin — COMMENTED** [review]
looks great, and thanks for splitting that method up 👍

File: code/edge/src/main/java/com/adobe/marketing/mobile/EdgeRequest.java [hunk 3]
```diff
@@ -77,6 +77,18 @@ JSONObject asJsonObject(final List<Map<String, Object>> serializedEvents) {
 			Utils.putIfNotEmpty(requestPayload, JSON_KEY_META, metadata.toObjectMap());
 		}
 
-		return new JSONObject(requestPayload);
+		JSONObject jsonPayload = null;
+		try {
+			jsonPayload = new JSONObject(requestPayload);
+		} catch (NullPointerException e) {
```
[the remark this answers is not in the document]
Comment (kevinlind) [reply code/edge/src/main/java/com/adobe/marketing/mobile/EdgeRequest.java:83]: Not that I'm seeing. From the public API, the data is passed through the EventDataUtil.clone method which checks for and removes any null keys. As JSONObject declares it throws a NullPointerException, I thought it prudent to add the try/catch.

**Review by emdobrin — APPROVED** [review]

File: code/edge/src/test/java/com/adobe/marketing/mobile/EdgeNetworkServiceTest.java [hunk 4]
```diff
@@ -644,6 +644,33 @@ public void testHandleStreamingResponse_SimpleStreamingResponse3() {
 		assertEquals(expectedResponses, result.onResponseCallback);
 	}
 
+	@Test
+	public void testHandleStreamingResponse_catchesIndexOutOfBoundsException() {
+		// When processing a response, the response is parsed into chunks using the line feed delimiter.
+		// Each chunk then removes the record separator to produce the expected response.
+		// If a chunk length is less than than record separator length, an IndexOutOfBounds exception is thrown.
```
Comment (emdobrin) [inline code/edge/src/test/java/com/adobe/marketing/mobile/EdgeNetworkServiceTest.java:651]: typo: than than

File: code/edge/src/main/java/com/adobe/marketing/mobile/CompletionCallbacksManager.java [hunk 5]
```diff
@@ -91,7 +91,17 @@ void unregisterCallback(final String requestEventId) {
 
 		if (callback != null) {
 			final List<EdgeEventHandle> handles = edgeEventHandles.get(requestEventId);
-			callback.onComplete(handles != null ? handles : new ArrayList<>());
+			try {
+				callback.onComplete(handles != null ? handles : new ArrayList<>());
+			} catch (Exception e) {
```
Comment (emdobrin) [inline code/edge/src/main/java/com/adobe/marketing/mobile/CompletionCallbacksManager.java:96]: we should not catch this exception and let it be handled on caller's side

File: code/edge/src/androidTest/java/com/adobe/marketing/mobile/CompletionHandlerFunctionalTests.java [hunk 6]
```diff
@@ -241,4 +241,47 @@ public void testSendEvent_withCompletionHandler_whenServerErrorAndHandle_callsCo
 		assertNetworkRequestCount();
 		assertTrue("Timeout waiting for EdgeCallback completion handler.", latch.await(1, TimeUnit.SECONDS));
 	}
+
+	@Test
+	public void testSendEvent_withCompletionHandler_whenExceptionThrownFromCallback_callsCompletionCorrectly()
```
Comment (emdobrin) [inline code/edge/src/androidTest/java/com/adobe/marketing/mobile/CompletionHandlerFunctionalTests.java:246]: see my comment about catching the callback exception, I would expect this test to call onComplete but throw the external extension

File: code/edge/src/main/java/com/adobe/marketing/mobile/EdgeHitProcessor.java [hunk 7]
```diff
@@ -276,6 +197,159 @@ public void onComplete() {
...
+				LOG_SOURCE,
+				"Failed to build the request payload, dropping current event (%s).",
+				entity.getEvent().getUniqueIdentifier()
+			);
+
+			return true; // Request complete, don't retry hit
+		}
+
+		Map<String, Object> edgeConfig = entity.getConfiguration();
+
```
Comment (addb) [inline code/edge/src/main/java/com/adobe/marketing/mobile/EdgeHitProcessor.java:235]: Should we add method to entity to get edgeConfigId directly? 

entity.getConfiguration.getEdgeConfigId()

File: code/edge/src/test/java/com/adobe/marketing/mobile/EdgeHitProcessorTests.java [hunk 8]
```diff
@@ -545,6 +545,30 @@ public void testProcessHit_consentUpdateEvent_emptyPayloadDueToInvalidData_doesN
 		assertProcessHit(entity, false, true);
 	}
 
+	@Test
+	public void testProcessHit_experienceEvent_noEdgeConfigId_doesNotSendNetworkRequest_returnsTrue() {
+		// Tests that when a good hit is processed that a network request is made and the request returns 200
+
+		// setup
+		Map<String, Object> config = new HashMap<>();
```
Comment (addb) [inline code/edge/src/test/java/com/adobe/marketing/mobile/EdgeHitProcessorTests.java:553]: Should we add test cases with null config map?

File: code/edge/src/test/java/com/adobe/marketing/mobile/EdgeNetworkServiceTest.java [hunk 9]
```diff
@@ -644,6 +644,33 @@ public void testHandleStreamingResponse_SimpleStreamingResponse3() {
...
+		// If a chunk length is less than the record separator length, an IndexOutOfBounds exception is thrown.
+		// This test verifies an IndexOutOfBounds exception is caught and handled.
+
+		networkService = new EdgeNetworkService(mockNetworkService);
+		final String recordSeparator = "<RS>";
+		final String lineFeed = "<LF>";
+		final String responseStr =
+			"<RS>{\"some\":\"thing1\\n\"}<LF>" +
+			"<RS>{\"some\":\"thing2\\n\"}<LF>" +
+			"<RS>{\"some\":\"thing3\\n\"}<LF>."; // note the trailing '.' character
```
Comment (addb) [inline code/edge/src/test/java/com/adobe/marketing/mobile/EdgeNetworkServiceTest.java:660]: Should we add comment saying "." is dropped with the exception?

File: code/edge/src/main/java/com/adobe/marketing/mobile/EdgeNetworkService.java [hunk 10]
```diff
@@ -412,7 +412,18 @@ void handleStreamingResponse(
 
 		while (scanner.hasNext()) {
 			final String jsonResult = scanner.next();
-			responseCallback.onResponse(jsonResult.substring(trimLength));
+			try {
```
Comment (addb) [inline code/edge/src/main/java/com/adobe/marketing/mobile/EdgeNetworkService.java:415]: Is it possible to avoid the exception all together by adding a check like this?
```
jsonResult.length > trimlength {
responseCallback.onResponse(jsonResult.substring(trimLength));
}
```

File: code/edge/src/main/java/com/adobe/marketing/mobile/CompletionCallbacksManager.java [hunk 11]
```diff
@@ -91,7 +93,21 @@ void unregisterCallback(final String requestEventId) {
...
+				StringWriter stringWriter = new StringWriter();
+				PrintWriter printWriter = new PrintWriter(stringWriter);
+				ex.printStackTrace(printWriter);
+
+				Log.warning(
+					LOG_TAG,
+					LOG_SOURCE,
+					"Exception thrown when invoking EdgeCallback.onComplete() for request event id %s: %s",
+					requestEventId,
+					stringWriter.toString()
```
Comment (emdobrin) [inline code/edge/src/main/java/com/adobe/marketing/mobile/CompletionCallbacksManager.java:108]: Reading more about this it looks like it is recommended to use the platform level logging support for exceptions instead of printStackTrace. 

@praveek any chance would could support for logging throwable in the Logging service as the platform provides?
https://developer.android.com/reference/android/util/Log#w(java.lang.String,%20java.lang.Throwable)

Alternatively we can look into using https://developer.android.com/reference/android/util/Log#getStackTraceString(java.lang.Throwable)

Comment (praveek) [reply code/edge/src/main/java/com/adobe/marketing/mobile/CompletionCallbacksManager.java:108]: @emdobrin We can explore about adding these additional log methods after the 2.0 release. For now, I would prefer [getStackTraceString](https://developer.android.com/reference/android/util/Log#getStackTraceString(java.lang.Throwable)) to get the string representation of stack trace. 

nit: Can we reword the log to `Exception thrown when invoking completion callback for request event id`
EdgeCallback.onComplete() is not very intuitive confusing when customer passes a lambda.

Comment (emdobrin) [reply code/edge/src/main/java/com/adobe/marketing/mobile/CompletionCallbacksManager.java:108]: @kevinlind can you make this update and also create a feature request for the logging service in aepsdk-core-android?

**Review by emdobrin — APPROVED** [review]
